### ISSUE TITLE
Auto Attach is not toggling for Node.js since latest update

### ISSUE THREAD

**Issue by rajsinghUSA** [microsoft/vscode#102834]
- VSCode Version: 1.47.2
- OS Version: Windows 10 v1909 (Build 18363.959)

Steps to Reproduce:

I use WSL to start VSCode with `code .` from my project directory in Windows Terminal. When VSCode starts up, "Auto Attach: Off" is shown in the status bar:

[![enter image description here][1]][1]

Clicking "Auto Attach: Off" in the status bar does not toggle it.

I then start my project in VSCode's integrated terminal with `npm run debug`, which corresponds to the following `scripts` line in `package.json`:
```js
"scripts": {
  "debug": "NODE_ENV=development DEBUG='knex:*' nodemon --inspect ./server.js"
}
```

However, the debugger is not auto-attached, so I need to manually run the `Attach` configuration from the *Run* panel shown on the left here:

[![enter image description here][2]][2]

The debugger works, but I still cannot toggle the `Auto Attach: Off` to be ***On*** instead:

[![enter image description here][3]][3]

Even if I **Ctrl+Shift+P** --> `Debug: Toggle Auto Attach`, nothing happens. Even if I restart VSCode after doing so.

----------

So, in conclusion, how can I start my VSCode with `code .` from Windows Terminal and then just type `npm run debug` and have VSCode, run my program AND auto-attach the debugger? I also need the server to restart upon file changes, which is why I am using `nodedemon` in the `npm run debug` script.

Do I need to either:

 - change `launch.json` configuration? Though I have tried essentially all of the boilerplate `launch.json` configs shown in VSCode, but none worked.

 - or modify my `package.json` defined `script` for `debug`?

 - or correct some setting on VSCode?

  [1]: https://i.stack.imgur.com/PFM6A.png
  [2]: https://i.stack.imgur.com/M2vTy.png
  [3]: https://i.stack.imgur.com/Cexao.png

Does this issue occur when all extensions are disabled?: Yes

**Comment by connor4312** [comment]
Can you please try the second set of instructions here? https://github.com/microsoft/vscode/issues/102057#issuecomment-657728467

**Comment by rajsinghUSA** [comment]
@connor4312 those instructions appear to be for linux (`echo $VSCODE_INSPECTOR_OPTIONS | awk '{if(match($0, /"([^"]+.deferred)"/,m))print m[1]}' | xargs rm`) what should I be doing on Windows?

**Comment by connor4312** [comment]
That step is not needed for Windows

**Comment by rajsinghUSA** [comment]
For settings `debug.node.autoAttach` it was already on "on":
![image](https://user-images.githubusercontent.com/17286177/87856694-bb65bd00-c8ee-11ea-8e8e-ea2104088af3.png)

So I then switched it to "off" and then "disabled". The "Auto Attach: Off" on the status bar has disappeared.

I should also mention I am noticing other debugging issues. For example, if I somehow get the debugger to run and attach (or perhaps launch), it won't actually break at a debugger unless I press `Ctrl+S` first (or somehow get it to reload at least once after starting it). I am not sure if this is a related issue or not, but if possible, I am fine with using the built-in `vscode-js-debug` if that can cleanly just get debugging working (as long as I can also get it to reload when files change, as `nodemon` does).

**Comment by connor4312** [comment]
> if I somehow get the debugger to run and attach (or perhaps launch), it won't actually break at a debugger statement unless I press Ctrl+S first (or somehow get it to reload at least once) after starting the npm script

It sounds like the debugger; statement might be executed before the attachment runs. The runtime will stop at `debugger;` statements without any particular behavior/instructions by the debugger (so long as a debugger is attached), so I would suspect something is happening with your build/run setup independent of js-debug.

**Comment by rajsinghUSA** [comment]
I don't have any build, just the straight `npm run debug`, which corresponds to:
```js
"scripts": {
  "debug": "NODE_ENV=development DEBUG='knex:*' nodemon --inspect ./server.js"
}
```

If it makes it easier by doing this from scratch with a blank `launch.json` and use a plain `package.json` script like `npm run start` corresponding to:
```js
"scripts": {
  "start": "node ./server.js"
}
```
what's the next step so I can get debugging working and breaking at my `debugger;` statements? And then what should I do to have it reload when files change?

After losing so many days to this issue, I would even be happy and eager to use VSCode's Live Share if that could help get to the bottom of the debugging issue.

**Comment by connor4312** [comment]
If you can collect a trace log using the 'launch' -- not attach -- scenario with the `debugger;` statement in place, I can look more into what you're hitting. Instructions:

> If you're able to, add `"trace": true` to your `launch.json` and reproduce the issue. The location of the log file on your disk will be written to the Debug Console. Share that with us.
>
> ⚠️ This log file will not contain source code, but will contain file paths. You can drop it into https://microsoft.github.io/vscode-pwa-analyzer/index.html to see what it contains. If you'd rather not share the log publicly, you can email it to connor@xbox.com

Also, if you're blocked on this you can temporarily use the old debugger by setting `debug.javascript.usePreview: false` in your user settings

**Comment by rajsinghUSA** [comment]
Debugging broke after VSCode did the update, I only turned the `debug.javascript.usePreview` on at a later time when trying to figure out a new setup that would work in the current version of VSCode. For example, my `launch.json` previously had `"type": "node"` and `"protocol": "inspector"`, neither of which are valid anymore.

I can certainly collect the trace log as you suggested, but before even getting to that part, what config should I have in my `launch.json` and what npm run-script line should I use in the `package.json`?

**Comment by rajsinghUSA** [comment]
Ok, here's an attempt with the following `launch.json` config (would still like some clarity on what may be a more appropriate `launch.json` config):

![image](https://user-images.githubusercontent.com/17286177/87865099-fe4f8100-c93e-11ea-818f-49df2c698068.png)

with `package.json` containing the following `start` script:
```js
  "scripts": {
    "start": "node ./server.js",
  }
```

I then click the "Start Debugging" run button:
![image](https://user-images.githubusercontent.com/17286177/87865139-70c06100-c93f-11ea-8cd2-fdba87310cc0.png)

Which runs this in the "Debug Console":
![image](https://user-images.githubusercontent.com/17286177/87865162-b67d2980-c93f-11ea-8467-62be5bbeb8e2.png)

And creates this log file so far:
[vscode-debugadapter-0.json.gz](https://github.com/microsoft/vscode/files/4943122/vscode-debugadapter-0.json.gz)

At this point the `debugger;` only works if it's immediately accessible. But there is no `debugger;` statement break if my frontend sends a request, even though my frontend gets the response just fine. In fact, the log appears to be exactly the same for before and after the frontend request.

I also can't get `debugger;` statements to break for frontend requests if I select "Node (preview)" from the debug panel's Run options, and then selecting `Debug: npm run start`:
![image](https://user-images.githubusercontent.com/17286177/87865625-cf88d900-c945-11ea-8c5e-8f0f33ac0b9a.png)
![image](https://user-images.githubusercontent.com/17286177/87865818-a406ee00-c947-11ea-85a8-87fb66adfd77.png)

Same happens if from the debug panel's Run options, I select `Debug: npm run debug` after "Node (preview)":
![image](https://user-images.githubusercontent.com/17286177/87865855-18da2800-c948-11ea-8607-886ecfea5211.png)

…then my frontend requests still don't break at the `debugger;` statement

Not sure what happened, but now I can't even do the weird trick where I press `Ctrl+S` and cause the server to restart once, and then `debugger;` statement in the request catches.

Again, in each of these cases, the frontend does get a proper response, just that the debugger isn't picking it up, and also I don't see any `console.log()` printouts in the console.

Sorry if this was extra long, just wanted to make sure to give you the full picture.

**Comment by connor4312** [comment]
Thank you for the details. Your launch.json with the npm script looks correct.

In your log, I don't see any issues that would make me suspicious, but I also don't see the debugger pausing anywhere. There are some cases where js-debug will automatically continue on a pause, which is what I thought _might_ have been happening here (though I hadn't seen it occur on a debugger; statement before) but it looks like that's not the case.

I wonder if maybe something in your framework is serving the request in a worker thread or something (which Node in general does not support debugging for at this point). Is the project you're working on open source? Maybe it would be easier for me to dig in a little bit myself.

**Comment by rajsinghUSA** [comment]
It's not open source but I am okay with adding you to the private github repo, I've lost many days at this point so really in a bind right now and willing to try whatever. Again, if you're okay with VSCode's Live Share, we can also try that, or even a regular screenshare app like Zoom or TeamViewer.

In the meantime I've added you to the private github repo, thank you.

**Comment by connor4312** [comment]
Thanks for the repo. I tried pulling it down and running it with the latest Node (14.5.0), but I think something isn't quite right. Just trying to `npm start` from a normal terminal, it tries to run server.js which is a module

```
➜  issue-102834 git:(master) ✗ node server.js 
(node:64322) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)
/Users/connor/Downloads/issue-102834/server.js:1
import express from "express";
^^^^^^

SyntaxError: Cannot use import statement outside a module

```

Correcting that,

```
➜  issue-102834 git:(master) ✗ node server.js
file:///Users/connor/Downloads/issue-102834/db/connection.js:3
import knexfile from "../knexfile.js";
       ^^^^^^^^
SyntaxError: The requested module '../knexfile.js' does not provide an export named 'default'
```

How are you actually building/running this project?

**Comment by rajsinghUSA** [comment]
Sorry @connor4312 I hadn't `push`ed my latest `commit` apparently; see the branch called `fixing_debugger` —that's exactly what I currently have. There is no build process at all, just using plain-Jane `express` in NodeJS with ES6 modules enabled in the `package.json`: `"type": "module"`

**Comment by connor4312** [comment]
Okay, great, thanks! I noticed you had debugger statement in /api/checkout/discount, so I

1. Npm installed'd
1. Uncommented the `hasPerm` middleware and tried hitting that debugger statement
1. F5'd
1. curl'd against that endpoint

<img width="1480" alt="image" src="https://user-images.githubusercontent.com/2230985/87957190-cfcbc600-ca64-11ea-83c6-8d22944a4fda.png">

Are there any addition steps needed here?

I wonder whether there's something happening on your network / some orphaned server stealing requests. E.g. if a nodemon server is listening on your IPv6 loopback and the "debug" server is listening on IPv4, or vise versa. Have you rebooted your computer since getting stuck here?

**Comment by Niffy** [comment]
I am having the same issue.

Before I  could just use `node --inspect-brk foo.js` and toggle auto attach to on, it would work straight away. 
Even the auto generated launch.json of attaching to node fails
```
{
  "configurations": [
  {
    "name": "Attach",
    "port": 9229,
    "request": "attach",
    "skipFiles": [
      "<node_internals>/**"
    ],
    "type": "pwa-node"
  }
  ]
}
```
with 
```
name": "Attach",
          ^

SyntaxError: Unexpected token ':'
    at wrapSafe (internal/modules/cjs/loader.js:1054:16)
    at Module._compile (internal/modules/cjs/loader.js:1102:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47
npm ERR! code ELIFECYCLE
```

In the terminal I cannot use node version 10, but can use version 12 (Both use the same version of npm 6.14.4)
```
nvm use 10
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module '"/Users/paulrobinson/Library/Application'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at Module._preloadModules (internal/modules/cjs/loader.js:901:12)
    at preloadModules (internal/bootstrap/node.js:602:7)
    at startup (internal/bootstrap/node.js:273:9)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
nvm is not compatible with the npm config "prefix" option: currently set to ""
Run `npm config delete prefix` or `nvm use --delete-prefix v10.20.1` to unset it.
```

This whole update has seriously affected my ability to debug, which is rather fustrating.

**Comment by connor4312** [comment]
@Niffy your issue is different, please follow the second set of instructions here to fix it: https://github.com/microsoft/vscode/issues/102057#issuecomment-657728467

**Comment by rajsinghUSA** [comment]
> Okay, great, thanks! I noticed you had debugger statement in /api/checkout/discount, so I
> 
> 1. Npm installed'd
> 2. Uncommented the `hasPerm` middleware and tried hitting that debugger statement
> 3. F5'd
> 4. curl'd against that endpoint
> > 
> Are there any additional steps needed here?
> 
> I wonder whether there's something happening on your network / some orphaned server stealing requests. E.g. if a nodemon server is listening on your IPv6 loopback and the "debug" server is listening on IPv4, or vise versa. Have you rebooted your computer since getting stuck here?

@connor4312 You are absolutely correct! After restarting the whole machine, `debugger;` statements began breaking again as expected. If I switch from `launch` to `attach` in `launch.json`, for example:
```js
    {
      "name": "Attach",
      "port": 9229,
      "request": "attach",
      "skipFiles": [
        "<node_internals>/**"
      ],
      "type": "pwa-node"
    },
```
It currently attaches fine if I manually press F5 or the "Start Debuggging" button, but how can I get auto-attach to work again, so I can just type `npm run debug` at the integrated terminal, and it automatically runs the debugger attached to it?

**Comment by connor4312** [comment]
Good to hear!

In 1.47.2 we reverted auto attach to its pre-1.47.0 behavior, so it should work identically as before.

Alternately, you can use the JavaScript debug terminal ("Create JavaScript debug terminal" command, or through the terminal switcher dropdown) and `npm start` inside of there.

**Comment by rajsinghUSA** [comment]
@connor4312 I see no option for "Auto Attach" in the status bar. I'm not sure how it was previously there, or how I previously enabled auto-attachment. This is my `launch.json` config:
```js
    {
      "name": "Attach",
      "port": 9229,
      "request": "attach",
      "skipFiles": [
        "<node_internals>/**"
      ],
      "type": "pwa-node"
    },
```
Can you please advise how to enable auto-attachment —whatever the previous way was (which is also now the current way as you said)?

**Comment by connor4312** [comment]
If you have it set to "disabled", it doesn't appear in the status bar. You can use the "Toggle Auto Attach" command to turn it back on, or manually update the `debug.node.autoAttach` setting in your user settings.

**Comment by rajsinghUSA** [comment]
@connor4312 I believe I have done that here: 

![image](https://user-images.githubusercontent.com/17286177/88314573-20e8ed80-cce3-11ea-9bb0-1ef0b8b27bcd.png)

But still there is no auto-attach occurring when I type `npm run debug` in the integrated terminal, nor is there an "Auto Attach" shown in the status bar.

**Comment by connor4312** [comment]
There's the hint there "Also modified in: Remote" which indicates that that setting is being overridden with a setting specific to your remote WSL

**Comment by rajsinghUSA** [comment]
That's because I enabled it within both:

![image](https://user-images.githubusercontent.com/17286177/88316052-32cb9000-cce5-11ea-9a7f-4e014c398219.png)

I even restarted all VSCode instances after setting "Auto Attach" to "On", but that did not make auto-attaching work or show up in the status bar.

**Comment by connor4312** [comment]
Did the `@builtin debug-auto-launch` extension get disabled somehow? Any errors in the log if you run the `Open Developer Tools` command?

**Comment by rajsinghUSA** [comment]
If this is what you're refer to, then I believe it is "Enabled":

![image](https://user-images.githubusercontent.com/17286177/88320309-67424a80-cceb-11ea-9621-275b28277377.png)

No errors in the Chrome devtools that pops up after choosing "Developer: Open Webview Developer Tools" from `Ctrl+Shift+P`. However, devtools oddly no longer opens after it did the first time, even after exiting all instances of VSCode and reopening.

**Comment by connor4312** [comment]
I'm not entirely sure what the issue is then. @weinand have you seen this before? (https://github.com/microsoft/vscode/issues/102834#issuecomment-663115993)

**Comment by rajsinghUSA** [comment]
Not sure why this was closed, the latest update did not resolve the issues. I will enumerate the current problems with debugging:

1. No ability to "**Auto-Attach**", as explained above.

2. **Hotreloading** does not work whether using `attach` or `launch` with the following `"scripts"` line in `package.json`
```js
    "debug": "NODE_ENV=development DEBUG='knex:*' nodemon --inspect ./server.js"
```
And the following `launch.json`:
```json
{
  "configurations": [
    {
      "name": "Attach",
      "port": 9229,
      "request": "attach",
      "skipFiles": [
        "<node_internals>/**"
      ],
      "type": "pwa-node"
    },
    {
      "name": "Launch via NPM",
      "request": "launch",
      "restart": true,
      "trace": true,
      "runtimeArgs": [
        "run-script",
        "debug"
      ],
      "runtimeExecutable": "npm",
      "skipFiles": [
        "<node_internals>/**"
      ],
      "type": "pwa-node"
    }
  ]
}
```

3. The devtools window that opens after selecting `Developer: Open Webview Developer Tools` from `Ctrl+Shift+P` no longer shows, even after updating and restarting VSCode and Windows 10.

Work has been impractical with the manual stopping/starting of the server to pickup the file changes. Please help.

**Comment by connor4312** [comment]
Sorry, closed this accidentally with others.

1. Ping @weinand -- have you seen this before? Currently we're using the old auto attach behavior in VS Code.
2. Please open a separate issue for this with details :)
3. The webview developer tools will always no-op if there's no webview visible.

**Comment by rajsinghUSA** [comment]
~3.~

**Comment by msanandrea** [comment]
Same issue here! This launch.json used to work fine

```
        {
            "type": "node",
            "request": "launch",
            "name": "Debug Marbot Infra",
            "protocol": "inspector",
            "program": "${workspaceFolder}/apps/marbot-infra/src/main.ts",
            "outFiles": ["${workspaceFolder}/dist/apps/marbot-infra/main.js"],
            "sourceMaps": true
        }
```

After an update (don't know which one) I'm getting 

```
import * as express from 'express';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```

**Comment by connor4312** [comment]
@msanandrea your problem is https://github.com/microsoft/vscode/issues/102873#issuecomment-674999289

**Comment by msanandrea** [comment]
> @msanandrea your problem is [#102873 (comment)](https://github.com/microsoft/vscode/issues/102873#issuecomment-674999289)

Thanks! This works!

```
 {
            "type": "node",
            "request": "launch",
            "name": "Debug launcher",
            "protocol": "inspector",
            "program": "${workspaceFolder}/apps/launcher/src/main.ts",
            "outFiles": ["${workspaceFolder}/dist/**/*.js"],
            "sourceMaps": true
}
```

**Comment by weinand** [comment]
@msanandrea BTW, the "protocol" and "sourceMaps" properties are no longer necessary...

**Comment by rajsinghUSA** [comment]
Hi @weinand, this still hasn't resolved with the latest update 1.48.2. I still have no ability to auto-attach. I'm happy to share my screen, even give you control, as I imagine debugging should be far easier that way.  It's been months like this, please help.

**Comment by connor4312** [comment]
The new version of VS Code (1.49) contains an entirely new version of auto attach -- a better version of what we briefly had in the 1.46 update. Give us a shout if it works better for you.

**Comment by msanandrea** [comment]
Hi @weinand , this is happening again after last update! version 1.50.1

```
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
    {
      "type": "node",
      "request": "launch",
      "name": "Debug Web Utils API",
      "program": "${workspaceFolder}/apps/web-utils-api/src/main.ts",
      "outFiles": ["${workspaceFolder}/dist/apps/**/*.js"]
    }
  ]
}

import * as express from 'express';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```

**Comment by connor4312** [comment]
Make sure your code is compiled to your outFiles.

Closing this issue as auto attach uses a new strategy and we've not heard back from the submitted whether they still have an issue.
